packages/core/src/RepoSync.ts

```diff
@@ -202,255 +163,96 @@ export class RepoSync {
-
-    // Record the doc's initial clock
-    await this.updateClock(documentId, ours, clock)
-  }
-
-  /**
-   * Checks the local doc's clock to ensure that it can be synced
-   * @param documentId
-   * @param clock
-   */
-  private validateDoc(documentId: string, clock: Clock) {
-    this.log('validateDoc', documentId)
-
-    // Make sure doc has a clock (i.e. is an Automerge object)
-    if (!clock) throw new TypeError(ERR_NOCLOCK)
-
-    // Make sure the document is newer than what we already have
-    const ourClock = this.getOurClock(documentId)
-    if (isMoreRecent(ourClock, clock)) throw new RangeError(ERR_OLDCLOCK)
-  }
+  // PRIVATE METHODS
 
-  /**
-   * Event listener that fires when any document is modified on the repo
-   * @param documentId
-   */
+  /** Event listener that fires when any document is modified on the repo */
   private async onDocChanged(documentId: string) {
-    this.log('onDocChanged', documentId)
-    const clock = await this.getClockFromDoc(documentId)
-
-    // make sure we can sync the new document
-    this.validateDoc(documentId, clock)
-
     // send the document if peer doesn't have it or has an older version
     await this.maybeSendChanges(documentId)
-
-    // see if peer has a newer version
-    await this.maybeRequestChanges(documentId, clock)
-
-    // update our clock
-    this.updateClock(documentId, ours, clock)
   }
 
-  /**
-   * Sends a hello message including our document count
-   */
+  /** Sends a hello message including our document count */
   private async sendHello() {
     const documentCount = this.repo.count
     this.log('sending hello', documentCount)
     this.send({ type: message.HELLO, documentCount })
   }
 
-  /**
-   * Checks whether peer has more recent information than we do; if so, requests changes
-   * @param documentId
-   * @param theirClock
-   */
-  private async maybeRequestChanges(
-    documentId: string,
-    theirClock: Clock | Promise<Clock> = this.getClockFromDoc(documentId)
-  ) {
-    this.log('maybeRequestChanges', documentId)
-    const ourClock = this.getOurClock(documentId)
-    theirClock = await theirClock
-    if (isMoreRecent(theirClock, ourClock)) this.advertise(documentId, theirClock)
-  }
-
-  /**
-   * Checks whether we have more recent information than they do; if so, sends changes
-   * @param documentId
-   */
+  /** Checks whether we have more recent information than they do; if so, sends changes */
   private async maybeSendChanges(documentId: string) {
     const theirClock = this.getTheirClock(documentId)
-    if (theirClock === undefined) return
-    const ourDoc = await this.repo.get(documentId)
-    const changes = _A.getMissingChanges(ourDoc, (theirClock as unknown) as A.Clock)
-    if (changes.length > 0) await this.sendChanges(documentId, changes)
+    const ourClock = this.getOurClock(documentId)
+    if (isMoreRecent(ourClock, theirClock)) {
+      const ourDoc = await this.repo.get(documentId)
```

Comment:
we can avoid unnecessarily pulling up the document by just comparing clocks first, and only calculating changes if we're sure we'll have them.